risc0/zkp/rust/src/field/goldilocks.rs

```diff
@@ -0,0 +1,745 @@
+        assert_eq!(fa + fb, Elem::from(a.wrapping_add(b)));
+    }
+
+    #[test]
+    fn compare_native_add_subfield_wrap() {
+        // Test for addition when P < a + b < 2^64
+        let fa = Elem::from(P - 2u64);
+        let fb = Elem::from(5u64);
+        let a: u64 = fa.into();
+        let b: u64 = fb.into();
+        assert_eq!(fa + fb, Elem::from(a.wrapping_add(b)));
+    }
+
+    #[test]
+    fn compare_native_add_u64_wrap() {
+        // Test for addition when P < a + b < 2^64
+        let fa = Elem::from(P - 2u64);
+        let fb = Elem::from(5u64);
+        let a: u64 = fa.into();
+        let b: u64 = fb.into();
+        assert_eq!(fa + fb, Elem::from(a.wrapping_add(b)));
+    }
+
+    #[test]
+    fn compare_native_subtraction_easy() {
+        // 0 < a - b < P  (when a > b)
+        let fa: Elem = Elem::from(P - 2u64);
+        let fb: Elem = Elem::from(3u64);
+        let a: u64 = fa.into();
+        let b: u64 = fb.into();
+        assert_eq!(
+            fa - fb,
+            Elem::from(((a as u128 - b as u128) % (P as u128)) as u64),
+        );
+    }
+
+    #[test]
+    fn compare_native_subtraction_negative_result() {
+        // Checks subtraction for `a-b` when P < (a - b) mod 2^64 < 2^64
+        // (when b < a, but a - b doesn't wrap all the way into our field)
+        let fa: Elem = Elem::from(1u64);
+        let fb: Elem = Elem::from(2u64);
+        let a: u64 = fa.into();
+        let b: u64 = fb.into();
+        assert_eq!(
+            fa - fb,
+            Elem::from(((a as u128 + (P - b) as u128) % (P as u128)) as u64),
+            "Error subtracting `{} - {}`",
+            a,
+            b
+        );
+    }
+
+    #[test]
+    fn compare_native_subtraction_prime_boundary() {
+        // Unlike baby bear, in which addition can't overflow a u64,
+        // here A + (P - B) is greater than 2^64,
+        // so this checks which modulus is being respected when
+        // P < (a - b) mod 2^64 < 2^64
+        let fa: Elem = Elem::from(P - 1u64);
+        let fb: Elem = Elem::from(P / 2u64);
+        let a: u64 = fa.into();
+        let b: u64 = fb.into();
+        assert_eq!(
+            fa - fb,
+            Elem::from(((a as u128 + (P - b) as u128) % (P as u128)) as u64),
+            "Error subtracting `{} - {}`",
+            a,
+            b
+        );
+    }
+
+    #[test]
+    fn compare_native_multiplication_easy() {
+        // 0 < a * b < P
+        // Simple case of o u64 overflow
+        let fa = Elem::from(2u64);
+        let fb = Elem::from(3u64);
+        let a: u64 = fa.into();
+        let b: u64 = fb.into();
+        assert_eq!(
+            fa * fb,
+            Elem::from(((a as u128 * b as u128) % P as u128) as u64),
+            "Error multiplying `{} * {}`",
+            a,
+            b
+        );
+    }
+
+    #[test]
+    fn compare_native_multiplication_subfield_wrap() {
+        // P < a * b < 2^64
+        // We expect the result mod P, but there's no u64 overflow
```

Comment:
I think this does overflow u64, as (3/2) P > 2^64 - 1. Which is a reasonable test to have, it's just not what the comment says.